### PR TITLE
Review

### PR DESCRIPTION
**Pull request by cartwrightforeman**
form for review

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+-AWS_SECRET_ACCESS_KEY=
+-S3_BUCKET=
+
+-AWS_ACCESS_KEY_ID=
+-AWS_SECRET_ACCESS_KEY=
+-S3_BUCKET=
+
+
+-SENDGRID_USERNAME=
+-SENDGRID_PASSWORD=
```
Comment (enthusiastick) [inline .env.example]: Hallelujah

File: app/controllers/api/v1/cities_controller.rb [hunk 2]
```diff
@@ -1,13 +1,38 @@
...
+    end
+  end
+
+  # private
+  #
+  # def city_params
+  #   params.require(:city).permit(:city, :body, :latitude, :longitude, :population, :state_id)
+  #   # params.fetch(:city, {}).permit(:city, :body, :latitude, :longitude, :population, :state_id)
+  # end
 end
```
Comment (enthusiastick) [inline app/controllers/api/v1/cities_controller.rb]: I am not super-excited to see all this commented-out code (especially as it includes your `strong_params`.)

File: app/controllers/api/v1/cities_controller.rb [hunk 3]
```diff
@@ -1,13 +1,38 @@
...
   end
+
+  def create
+    @state = State.find(params[:state_id])
+    data = JSON.parse(request.body.read)
+    @city = @state.cities.new(data['city'])
+    if @city.save
+      render json: data['city']
+    else
+      redirect_to state_cities_path
```
Comment (enthusiastick) [inline app/controllers/api/v1/cities_controller.rb]: Disappointed.

Comment (enthusiastick) [inline app/controllers/api/v1/cities_controller.rb]: We do not, as a rule, `redirect` in our API controllers. API endpoints are for `fetch` requests; they should send and receive JSON. If an action fails, then they should serve errors as JSON (for the front-end to consume.)

File: app/controllers/api/v1/reviews_controller.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  def create
+    @state = State.find(params[:state_id])
+    @city = @state.cities.find(params[:city_id])
+    data = JSON.parse(request.body.read)
+    @review = @city.reviews.new(data['review'])
+    @review.user_id = current_user.id
+    if @review.save
+      render json: data["review"]
+    else
+      redirect_to new_state_city_path(@city.id)
```
Comment (enthusiastick) [inline app/controllers/api/v1/reviews_controller.rb]: Not best practice.

See my above comments regarding redirects in API controllers.

File: app/controllers/api/v1/cities_controller.rb (hunk 3 above)
Comment (cartwrightforeman) [inline app/controllers/api/v1/cities_controller.rb]: We tried using strong params and it wouldn't work, so we received help from Chrissy and Steven on this.

File: react/src/containers/ReviewShowContainer.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      method: 'POST',
+      credentials: 'same-origin',
+      body: JSON.stringify(formPayload)
+    })
+    .then(response => response.json())
+    .then(responseData => {
+      console.log('IT WORKED for REVIEWS')
+      this.setState({ reviews: [...this.state.reviews, responseData] })
+    })
+  }
```
Comment (enthusiastick) [inline react/src/containers/ReviewShowContainer.js]: Again, as a best practice we don't merge comments to `master`, and try to clean up these debugging statements (e.g. `console.log()`. Also, as there is no `.catch()`, there is no error-handling here if things do go wrong (and, as stated, the redirect in your controller will have no effect.)

Comment (cartwrightforeman) [inline react/src/containers/ReviewShowContainer.js]: Oh okay, now I see what you're saying. Working on that now!

File: app/controllers/api/v1/cities_controller.rb [hunk 6]
```diff
@@ -1,13 +1,29 @@
...
+    @city = City.new
+  end
+
+  def create
+    @state = State.find(params[:state_id])
+    data = JSON.parse(request.body.read)
+    @city = @state.cities.new(data['city'])
+    if @city.save
+      render json: data['city']
+    end
```
Comment (enthusiastick) [inline app/controllers/api/v1/cities_controller.rb]: This is *better*. But what would be even better would be to leverage the `ActiveRecord` errors, e.g.:

```Ruby
  if @city.save
    render json: data['city']
  else
    render json: { errors: @city.errors }, status: :unprocessable_entity
  end
```

**Comment by enthusiastick** [comment]
I'll allow it.

**Pull request merged**